openvdb/tools/FastSweeping.h

```diff
@@ -0,0 +1,1432 @@
+
+template <typename GridT>
+FastSweeping<GridT>::FastSweeping()
+    : mGrid1(nullptr), mGrid2(nullptr), mCoords(nullptr), mPagedArray(), mBoundaryCount(0)
+{
+}
+
+template <typename GridT>
+void FastSweeping<GridT>::clear()
+{
+    mGrid1.reset();
+    mGrid2.reset();
+    mCoords.reset();
+    mPagedArray.clear();
+    mBoundaryCount = 0;
+}
+
+template <typename GridT>
+void FastSweeping<GridT>::initSdf(const GridT &fogGrid, ValueT isoValue, bool isInputSdf)
+{
+    this->clear();
+    mGrid1 = fogGrid.deepCopy();//very fast
+    InitSdf kernel(*this);
+    kernel.run(isoValue, isInputSdf);
+}
+
+template <typename GridT>
+template <typename OpT>
+void FastSweeping<GridT>::initExt(const GridT &fogGrid, const OpT &op, ValueT isoValue, bool isInputSdf)
+{
+    this->clear();
+    mGrid1 = fogGrid.deepCopy();//very fast
+    mGrid2 = fogGrid.deepCopy();//very fast
+    InitExt<OpT> kernel(*this);
+    kernel.run(isoValue, op, isInputSdf);
+}
+
+template <typename GridT>
+void FastSweeping<GridT>::initDilate(const GridT &sdfGrid, int dilate, NearestNeighbors nn)
+{
+    this->clear();
+    mGrid1 = sdfGrid.deepCopy();//very fast
+    DilateKernel kernel(*this);
+    kernel.run(dilate, nn);
+}
+
+template <typename GridT>
+template<typename MaskTreeT>
+void FastSweeping<GridT>::initMask(const GridT &sdfGrid, const Grid<MaskTreeT> &mask, bool ignoreActiveTiles)
+{
+    this->clear();
+    mGrid1 = sdfGrid.deepCopy();//very fast
+
+    if (mGrid1->transform() != mask.transform()) {
+        OPENVDB_THROW(RuntimeError, "FastSweeping: Mask not aligned with the grid!");
+    }
+
+    if (mask.getGridClass() == GRID_LEVEL_SET) {
+        using T = typename MaskTreeT::template ValueConverter<bool>::Type;
+        typename Grid<T>::Ptr tmp = sdfInteriorMask(mask);//might have active tiles
+        tmp->tree().voxelizeActiveTiles();//multi-threaded
+        MaskKernel<T> kernel(*this);
+        kernel.run(tmp->tree());//multi-threaded
+    } else {
+        if (ignoreActiveTiles || !mask.tree().hasActiveTiles()) {
+            MaskKernel<MaskTreeT> kernel(*this);
+            kernel.run(mask.tree());//multi-threaded
+        } else {
+            using T = typename MaskTreeT::template ValueConverter<ValueMask>::Type;
+            T tmp(mask.tree(), false, TopologyCopy());//multi-threaded
+            tmp.voxelizeActiveTiles(true);//multi-threaded
+            MaskKernel<T> kernel(*this);
+            kernel.run(tmp);//multi-threaded
+        }
+    }
+}// FastSweeping::initMaskSdf
+
+template <typename GridT>
+void FastSweeping<GridT>::sweep(int nIter, bool finalize)
+{
+    if (!mGrid1) {
+      OPENVDB_THROW(RuntimeError, "FastSweeping::sweep called before initialization");
+    }
+    if (this->boundaryCount() == 0) {
+        OPENVDB_THROW(RuntimeError, "FastSweeping: No boundary voxels found!");
+    } else if (this->voxelCount() == 0) {
+        OPENVDB_THROW(RuntimeError, "FastSweeping: No computing voxels found!");
```

Comment:
I'm running into this runtime error consistently with different inputs. Although I've figured out what it means it's not immediately clear. For example, calling `dilateSdf` with a narrow band level set or even with a filled but small interior sdf will cause this check to fire. Is there a way we can handle this case more garcefully?

Reply:
yes probably but we should discuss what behaviour you would expect/refer since it's not intuitive (at least not to me)

Reply:
Did we decide on a plan for this runtime error? Are we going to solve this at the SOP level? My concern is what users will see in Houdini when this gets hit as this won't mean anything to most people!

Reply:
this is no longer and issue since all free-standing functions are careful never to call FastSweeping::sweep if and of the initialization methods failed